iree/compiler/Dialect/Flow/Transforms/ReplaceFuncArgsWithConstants.cpp

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+namespace {
+
+// Replace the function arguments with constants if it has "iree.module.export"
+// attribute.
+class FuncArgsReplacementPattern : public OpRewritePattern<FuncOp> {
+  using OpRewritePattern<FuncOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(FuncOp funcOp,
+                                PatternRewriter &rewriter) const override {
+    if (!funcOp.getAttr("iree.module.export")) {
+      return failure();
+    }
+    if (funcOp.getNumArguments() == 0) {
+      return failure();
+    }
+    Location loc = funcOp.getLoc();
+    auto funcType = rewriter.getFunctionType({}, funcOp.getType().getResults());
+    auto newFuncOp = rewriter.create<FuncOp>(loc, funcOp.getName(), funcType);
+    newFuncOp.setAttr("iree.module.export", rewriter.getUnitAttr());
```

Comment:
Set other function attributes here?

Reply:
We don't want to set reflection attributes here. The later pass will look into arguments and set it I think.

Reply:
Not attributes added during Flow dialect conversion. IIRC bert mail entry mhlo saved model does have some compute related attribute.

Reply:
I think it should copy the attributes from the existing function.

Reply:
There are two extra attributes, one is iree.reflection and another is tf.entry_function. The latter one is not needed because we don't connect to tf at this level.

The former one can be added later in MaterializeExportedReflectionPass. I'm not pretty sure what happens if there are some attributes already. I'm sure that this will add correct arguments/results reflection even they don't exist, which is the most of cases in IREE e2e tests.

I also tested with the real mobile bert model, and it works with the rewritten func op. eg,

iree-benchmark-module --module_file=bert.vmfb --driver=vmla --entry_function=serving_default

Reply:
But yes, I think we can keep other attributes. I just feel that they are not important in IREE...

Reply:
Yeah, neither of those should be required. tf* stuff is not required at all and the reflection stuff is only needed if you are calling from pyiree.

Reply:
I think the principal of preserving attributes in general is still good though. If you're trying to just clone an op then preserving its attributes should be the default IMO. If there are some specific attributes we want to drop (like all attributes from a TF dialect) we should add a separate pass to do that. That said, the current infra makes it too easy to drop attributes in all kinds of passes, so probably we do it all over the place already :-/

Reply:
I agree that preserving attributes in general is good. However, in this case, we create a new func for benchmarking purpose. So I think it is okay to drop all the attributes for "created function", and just add needed attributes for this. It's not the case of replacing old functions now.